Google.Api.Gax.Grpc.Tests/AsyncResponseStreamTest.cs

```diff
@@ -0,0 +1,109 @@
+﻿/*
+ * Copyright 2019 Google LLC
+ * Use of this source code is governed by a BSD-style
+ * license that can be found in the LICENSE file or at
+ * https://developers.google.com/open-source/licenses/bsd
+ */
+
+using Google.Api.Gax.Grpc.Testing;
+using Grpc.Core;
+using Moq;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Api.Gax.Grpc.Tests
+{
+    public class AsyncResponseStreamTest
+    {
+        [Fact]
+        public void GetAsyncEnumerator_ReturnsSelf()
+        {
+            var stream = CreateStream(1, 2, 3);
+            Assert.Same(stream, stream.GetAsyncEnumerator());
+        }
+
+        [Fact]
+        public void GetAsyncEnumerator_FailsSecondCall()
+        {
+            var stream = CreateStream(1, 2, 3);
+            stream.GetAsyncEnumerator();
+            Assert.Throws<InvalidOperationException>(() => stream.GetAsyncEnumerator());
+        }
+
+        // Tests for the data
+
+        [Fact]
+        public async Task MoveNextAsync_Parameterless()
+        {
+            var stream = CreateStream(1, 2);
+            Assert.True(await stream.MoveNextAsync());
+            Assert.Equal(1, stream.Current);
+            Assert.True(await stream.MoveNextAsync());
+            Assert.Equal(2, stream.Current);
+            Assert.False(await stream.MoveNextAsync());
+        }
+
+        [Fact]
+        public async Task MoveNextAsync_WithCancellationToken()
+        {
+            var stream = CreateStream(1, 2);
+            Assert.True(await stream.MoveNextAsync(default));
+            Assert.Equal(1, stream.Current);
+            Assert.True(await stream.MoveNextAsync(default));
+            Assert.Equal(2, stream.Current);
+            Assert.False(await stream.MoveNextAsync(default));
+        }
+
+        // Tests for cancellation tokens being propagated
+
+        [Fact]
+        public void MoveNextAsync_Parameterless_WithoutGetAsyncEnumerator_UsesCancellationTokenNone()
+        {
+            var mock = new Mock<IAsyncStreamReader<int>>(MockBehavior.Strict);
+            mock.Setup(x => x.MoveNext(CancellationToken.None)).Returns(Task.FromResult(true)).Verifiable();
+
+            var stream = new AsyncResponseStream<int>(mock.Object);
+            Assert.True(stream.MoveNextAsync().Result);
+
+            mock.Verify();
+        }
+
+        [Fact]
+        public void MoveNextAsync_Parameterless_AfterGetAsyncEnumerator_RespectsToken()
```

Comment:
nit: the RespectsToken part of the name threw me off, I would associate that more with the token being canceled and that cancelation respected. Consider changing to PropagatesToken, this one and the following.

Reply:
Works for me.